pkg/distributor/ha_tracker.go

```diff
@@ -0,0 +1,291 @@
+package distributor
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/weaveworks/common/mtime"
+
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/ring"
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+const (
+	longPollDuration = 10 * time.Second
+)
+
+// Track the most recent timestamp for the instance we're
+// accepting samples from in a Prometheus HA setup.
+type instanceTracker struct {
+	instance  string
+	timestamp int64
+}
+
+// Track the instance we're accepting samples from
+// for each HA cluster we know about.
+type tracker struct {
+	logger log.Logger
+
+	// Instances we are accepting samples from.
+	electedLock sync.RWMutex
+	elected     map[string]instanceTracker
+	done        chan struct{}
+	quit        context.CancelFunc
+	// The labels to look for as bytes.
+	clusterBytes []byte
+	replicaBytes []byte
+	prefix       string
+
+	// We should only update the timestamp this often
+	writeTimeout int64
+
+	consulLock       sync.RWMutex
+	overwriteTimeout int64 // Timeout that has to occur before we overwrite to a different instance
+	client           ring.KVClient
+}
+
+// Config contains the configuration require to
+// create a Distributor
+type HATrackerConfig struct {
+	ReplicaLabel string
+	ClusterLabel string
+	Prefix       string
+
+	consul ring.ConsulConfig
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *HATrackerConfig) RegisterFlags(f *flag.FlagSet) {
+	f.StringVar(&cfg.ReplicaLabel, "ha-tracker.replica", "replica", "Prometheus label to look for in samples to identify a Proemtheus HA replica.")
+	f.StringVar(&cfg.ClusterLabel, "ha-tracker.cluster", "cluster", "Prometheus label to look for in samples to identify a Poemtheus HA cluster.")
+	f.StringVar(&cfg.Prefix, "ha-tracker.prefix", "prom_ha", "Prefix to store HA cluster accept instance information under within KV store.")
+	// Need to copy these flags since we don't want to break compatability of the existing 'consul.<var>' flags
+	// but we want the ability to use different Consul instances for the ring and for HA cluster tracking.
+	f.StringVar(&cfg.consul.Host, "ha-tracker.consul-hostname", "localhost:8500", "Hostname and port of Consul.")
+	f.StringVar(&cfg.consul.ACLToken, "ha-tracker.consul-acltoken", "", "ACL Token used to interact with Consul.")
+	f.DurationVar(&cfg.consul.HTTPClientTimeout, "ha-tracker.consul-client-timeout", 2*longPollDuration, "HTTP timeout when talking to consul")
+	f.BoolVar(&cfg.consul.ConsistentReads, "ha-tracker.consul-consistent-reads", true, "Enable consistent reads to consul.")
```

Comment:
You should use `cfg.consul.RegisterFlags()` and update it to take an (optional) prefix.